bitrise/util.go

```diff
@@ -49,12 +31,119 @@ func ReadBitriseConfigYML(pth string) (models.BitriseConfigModel, error) {
+func ReadSpecStep(pth string) (models.StepModel, error) {
+	if isExists, err := pathutil.IsPathExists(pth); err != nil {
+		return models.StepModel{}, err
+	} else if !isExists {
+		return models.StepModel{}, NewErrorf("No file found at path", pth)
+	}
+
+	bytes, err := ioutil.ReadFile(pth)
+	if err != nil {
+		return models.StepModel{}, err
+	}
+	var specStep models.StepModel
+	if err := yaml.Unmarshal(bytes, &specStep); err != nil {
+		return models.StepModel{}, err
+	}
+
+	return specStep, nil
+}
+
+// MergeSpecStep ...
+func MergeSpecStep(specStep, workflowStep models.StepModel) models.StepModel {
+	specStep.ID = mergeString(specStep.ID, workflowStep.ID)
+	specStep.SteplibSource = mergeString(specStep.SteplibSource, workflowStep.SteplibSource)
+	specStep.VersionTag = mergeString(specStep.VersionTag, workflowStep.VersionTag)
+	specStep.Name = mergeString(specStep.Name, workflowStep.Name)
+	specStep.Description = mergeString(specStep.Description, workflowStep.Description)
+	specStep.Website = mergeString(specStep.Website, workflowStep.Website)
+	specStep.ForkURL = mergeString(specStep.ForkURL, workflowStep.ForkURL)
+	specStep.Source = mergeStringStringMap(specStep.Source, workflowStep.Source)
+	specStep.HostOsTags = mergeStringSlice(specStep.HostOsTags, workflowStep.HostOsTags)
+	specStep.ProjectTypeTags = mergeStringSlice(specStep.ProjectTypeTags, workflowStep.ProjectTypeTags)
+	specStep.TypeTags = mergeStringSlice(specStep.TypeTags, workflowStep.TypeTags)
+	specStep.IsRequiresAdminUser = workflowStep.IsRequiresAdminUser
+	specStep.Inputs = mergeInputModels(specStep.Inputs, workflowStep.Inputs)
+	specStep.Outputs = mergeInputModels(specStep.Outputs, workflowStep.Outputs)
+	return specStep
+}
+
+func mergeBoolPtr(reference, override *bool) *bool {
+	if override != nil {
+		return override
+	}
+	return reference
+}
+
+func mergeInputModel(reference, override models.InputModel) models.InputModel {
+	reference.MappedTo = mergeString(reference.MappedTo, override.MappedTo)
+	reference.Title = mergeString(reference.Title, override.Title)
+	reference.Description = mergeString(reference.Description, override.Description)
+	reference.Value = mergeString(reference.Value, override.Value)
+	reference.ValueOptions = mergeStringSlice(reference.ValueOptions, override.ValueOptions)
+	reference.IsRequired = mergeBoolPtr(reference.IsRequired, override.IsRequired)
+	reference.IsExpand = mergeBoolPtr(reference.IsExpand, override.IsExpand)
+	reference.IsDontChangeValue = mergeBoolPtr(reference.IsDontChangeValue, override.IsDontChangeValue)
+	return reference
+}
+
+func mergeInputModels(reference, override []models.InputModel) []models.InputModel {
+	for idx, referenceInput := range reference {
+		for _, overrideInput := range override {
+			if referenceInput.MappedTo == overrideInput.MappedTo {
+				reference[idx] = mergeInputModel(referenceInput, overrideInput)
+			}
+		}
+	}
+	return reference
+}
+
+func mergeStringSlice(reference, override []string) []string {
```

Comment:
mergeStringSlice -> not required, if the `override` one exists just use that, no need for merge etc.